File: serving/runtime/src/telemetry.rs
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

//! Reusable serving counters with bounded names.

use mindclade_telemetry::CounterRegistry;
use std::collections::BTreeMap;

#[derive(Clone, Debug, Default)]
pub struct ServingMetrics {
    counters: CounterRegistry,
}

impl ServingMetrics {
    pub fn admitted(&self) {
        let _ = self.counters.add("serving.admitted", 1);
    }
    pub fn rejected(&self) {
        let _ = self.counters.add("serving.rejected", 1);
    }
    pub fn completed(&self) {
        let _ = self.counters.add("serving.completed", 1);
    }
    #[must_use]
    pub fn snapshot(&self) -> BTreeMap<String, u64> {
        self.counters.snapshot()
    }
}

File: serving/model_worker/telemetry.py
# Copyright © 2026 Mindclade, LLC. All Rights Reserved.
# Mindclade Proprietary and Confidential.
# SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
#

"""Low-cardinality model-worker counters."""

from dataclasses import dataclass
from threading import Lock


@dataclass(frozen=True, slots=True)
class WorkerMetrics:
    admitted: int
    completed: int
    rejected: int
    failed: int


class Telemetry:
    def __init__(self) -> None:
        self._values = [0, 0, 0, 0]
        self._lock = Lock()

    def increment(self, outcome: str) -> None:
        indexes = {"admitted": 0, "completed": 1, "rejected": 2, "failed": 3}
        if outcome not in indexes:
            raise ValueError("model-worker outcome is invalid")
        with self._lock:
            self._values[indexes[outcome]] += 1

    def snapshot(self) -> WorkerMetrics:
        with self._lock:
            return WorkerMetrics(*self._values)
